## Scenario

Mandisa Nkosi is actively pursuing a career in data analysis. 
She has established a relationship with a political party that needs to decide 
how best to invest its available advertising budget. Mandisa thinks she can 
explore potential advertising avenues by analysing films that are available 
on streaming platforms. She will use the movies_merge.xlsx and ott_merge.csv data sets.

The insights gained from the analysis will inform the campaign, 
promotional materials, slogans, and language the political party will use 
to reach potential voters. You’ll return to this analysis throughout the week. 

## Objective - PA 3.1.5

To prepare your files and Jupyter Notebook for analysis:

    1. Import the CSV files into DataFrames.
    2. View the DataFrames.
    3. Describe the DataFrames to understand the structures and data types. 
    4. Merge the two DataFrames into a single DataFrame.

In [1]:
# Import the library.
import pandas as pd

In [2]:
# Import the datasets.
movies_merge = pd.read_excel('movies_merge.xlsx')
ott_merge = pd.read_csv('ott_merge.csv')

In [6]:
# View the 'movies_merge' DataFrame.
print("Shape of the movies_merge df is, ",movies_merge.shape)
movies_merge.head()

Shape of the movies_merge df is,  (16744, 11)


ID                           Title  Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010  13+   8.8             0.87   
1   2                      The Matrix  1999  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018  13+   8.5             0.84   
3   4              Back to the Future  1985   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  
0  United States,United Kingdom  English,Japanese,French    148.0  
1                 United States                  English    136.0  
2                 United States                  English    149.0  
3                 United States                  English    116.0  
4      Italy,Spain,West Germany                  Italian    161.0

In [5]:
# View the 'ott_merge' DataFrame.
print("Shape of the ott_merge df is, ",ott_merge.shape)
ott_merge.head()

Shape of the ott_merge df is,  (16744, 5)


ID  Netflix  Hulu  Prime Video  Disney+
0   1        0     0            1        0
1   2        0     1            0        0
2   3        0     0            1        0
3   4        1     0            0        0
4   5        0     0            1        0

In [9]:
# Describe the DataFrame
print("The data types for movies_merge are:\n",movies_merge.dtypes)
print("\nThe data types for ott_merge are:\n",ott_merge.dtypes)

The data types for movies_merge are:
 ID                   int64
Title               object
Year                 int64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
dtype: object

The data types for ott_merge are:
 ID             int64
Netflix        int64
Hulu           int64
Prime Video    int64
Disney+        int64
dtype: object


In [11]:
# Merge the DataFrame.
merged_data = pd.merge(movies_merge, ott_merge,
                      on='ID', how='left')

# View the merged_data
print("Shape of the merged_data df is, ",merged_data.shape)
print("\nThe data types for merged_data are:\n",merged_data.dtypes)
merged_data.head()

Shape of the merged_data df is,  (16744, 15)

The data types for merged_data are:
 ID                   int64
Title               object
Year                 int64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
Netflix              int64
Hulu                 int64
Prime Video          int64
Disney+              int64
dtype: object


ID                           Title  Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010  13+   8.8             0.87   
1   2                      The Matrix  1999  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018  13+   8.5             0.84   
3   4              Back to the Future  1985   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  Netflix  \
0  United States,United Kingdom  English,Japanese,French    148.0        0   
1                 United States                  English    136.0        0   
2                 United States                  English    149.0        0   
3                 United States                  English    116.0        1   
4      Italy,Spain,West Germany                  Italian    161.0        0   

   Hulu  Prime Video  Disney+  
0     0            1        0  
1     1            0        0  
2     0            1        0  
3     0            0        0  
4     0            1        0

In [13]:
# Concatenate the DataFrame.
concat_data = pd.concat([movies_merge, ott_merge], axis=0)

# View the concat_data
print("Shape of the concat_data df is, ",concat_data.shape)
print("\nThe data types for merged_data are:\n",concat_data.dtypes)
concat_data.head()

Shape of the concat_data df is,  (33488, 15)

The data types for merged_data are:
 ID                   int64
Title               object
Year               float64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
Netflix            float64
Hulu               float64
Prime Video        float64
Disney+            float64
dtype: object


ID                           Title    Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010.0  13+   8.8             0.87   
1   2                      The Matrix  1999.0  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018.0  13+   8.5             0.84   
3   4              Back to the Future  1985.0   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966.0  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  Netflix  \
0  United States,United Kingdom  English,Japanese,French    148.0      NaN   
1                 United States                  English    136.0      NaN   
2                 United States                  English    149.0      NaN   
3                 United States                  English    116.0      NaN   
4      Italy,Spain,West Germany                  Italian    161.0      NaN   

   Hulu  Prime Video  Disney+  
0   NaN          NaN      NaN  
1   NaN          NaN      NaN  
2   NaN          NaN      NaN  
3   NaN          NaN      NaN  
4   NaN          NaN      NaN

When the concat function was used the duplicates were included in the final DataFrame,
so the total rows are 33484. But the merge funtion takes care of the duplicates and
so the total rows of the DataFrame is 16744. 

### Objective - PA 3.1.9 Use groupby() and aggregate() functions
Use the groupby() and aggregate() functions on the merged DataFrame to:

    1. Determine how many films from each year (released from 2012 to the present) 
       were watched on Netflix.
    2. Determine what the average runtime is of movies released each year.
    3. Determine what the best and worst reviews are that movies received 
       on Rotten Tomatoes.

### 1. Determine how many films from each year (released from 2012 to the present) were watched on Netflix. 

In [18]:
mo_gpby = merged_data.groupby('Year')[['Netflix']].agg('sum').reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year  Netflix
100  2012      195
101  2013      200
102  2014      202
103  2015      252
104  2016      236
105  2017      294
106  2018      307
107  2019      138
108  2020       31

### 2. Determine what the average runtime is of movies released each year since 2012.

In [23]:
mo_gpby = merged_data.groupby('Year')[['Runtime']].agg(['sum','mean']).reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year   Runtime           
                sum       mean
100  2012   73405.0  90.511714
101  2013   85687.0  91.545940
102  2014   87788.0  92.995763
103  2015   95795.0  92.914646
104  2016  107621.0  93.992140
105  2017  125822.0  94.460961
106  2018  112995.0  94.635678
107  2019   60997.0  93.410413
108  2020   12029.0  93.976562

### 3. Determine what the best and worst reviews are that movies received on Rotten Tomatoes since 2012.

In [24]:
mo_gpby = merged_data.groupby('Year')[['Rotten Tomatoes']].agg(['max','min']).reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year Rotten Tomatoes      
                      max   min
100  2012             1.0  0.04
101  2013             1.0  0.02
102  2014             1.0  0.05
103  2015             1.0  0.05
104  2016             1.0  0.02
105  2017             1.0  0.04
106  2018             1.0  0.06
107  2019             1.0  0.05
108  2020             1.0  0.06